### DISCUSSION TITLE
How to create multiple streams with custom ID? And how to pause/resume these streams?

### DISCUSSION THREAD

**Discussion by parthibx24** [alibaba/xquic#339] [Q&A] [answered]
Hello, I am reading the test client and trying to create a simple TRANSPORT alpn mode client program based on it. Right now I'm trying to understand how I can create multiple streams using a single quic conn. Looking at test client, to create a stream we do `user_stream->stream = xqc_stream_create(ctx.engine, cid, user_stream);`, which internally creates a stream with default Id. And it's not possible to change stream->stream_id. So, I'm not sure how it is possible to create more streams if every time xqc_stream_create stream is called it creates stream with the same default I'd.
Also, how can I pause these stream, after creating the stream the `stream write notify` callback gets called immediately, unless the body is empty it does not stop calling this function.

**Comment by Kulsk** [comment, marked as answer]
1. stream id is sequential and is not designed to be controlled by users. otherwise it is vulnerable as many QUIC implementations might create those streams with stream_id lower than the current. some QUIC implementations might consider this as STREAM_LIMIT_ERROR connection error. for example, the peer defines MAX_BIDI_STREAMS as 100 and expects stream starts with stream_id 0, and user come up with a stream with stream_id 404, which exceeds 100 and the peer might close connection.

2. if you got nothing to send, do nothing when stream_write_notify callback triggers, when you get some data to send, try xqc_stream_send. if the send window is full, you can continue sending the remain data when stream_write_notify triggers.
